src/ucs/datastruct/ptr_array.h

```diff
@@ -121,12 +145,228 @@ void *ucs_ptr_array_replace(ucs_ptr_array_t *ptr_array, unsigned index, void *ne
+ */
+
+
+/* Locked ptr array */
+typedef struct ucs_ptr_array_locked {
+    ucs_ptr_array_t          super;
+    ucs_recursive_spinlock_t lock;
+} ucs_ptr_array_locked_t;
+
+
+/**
+ * Locked array init
+ *
+ * @param [in] locked_ptr_array  Pointer to a locked ptr array.
+ * @param [in] name              The name of the ptr array.
+ *
+ * @return Success or failure.
+ *
+ */
+ucs_status_t
+ucs_ptr_array_locked_init(ucs_ptr_array_locked_t *locked_ptr_array,
+                          const char *name);
+
+
+/**
+ * Cleanup the locked array.
+ *
+ * @param [in] locked_ptr_array    Pointer to a locked ptr array.
+ *
+ * @note All values should already be removed from it.
+ */
+void ucs_ptr_array_locked_cleanup(ucs_ptr_array_locked_t *locked_ptr_array);
+
+
+/**
+ * Insert a pointer to the locked array.
+ *
+ * @param [in]  locked_ptr_array  Pointer to a locked ptr array.
+ * @param [in]  value             Pointer to insert. Must be 8-byte aligned.
+ *
+ * @return The index to which the value was inserted.
+ *
+ * Complexity: Amortized O(1)
+ *
+ * @note The array will grow if needed.
+ */
+unsigned ucs_ptr_array_locked_insert(ucs_ptr_array_locked_t *locked_ptr_array,
+                                     void *value);
 
 
+/**
+ * Remove a pointer from the locked array.
+ *
+ * @param [in] locked_ptr_array  Pointer to a locked ptr array.
+ * @param [in] index             Index to remove from.
+ *
+ * Complexity: O(1)
+ */
+void ucs_ptr_array_locked_remove(ucs_ptr_array_locked_t *locked_ptr_array,
+                                 unsigned index);
+
+
+/**
+ * Replace pointer in the locked array
+ *
+ * @param [in] locked_ptr_array  Pointer to a locked ptr array.
+ * @param [in] index             Index of slot
+ * @param [in] new_val           Value to put into slot given by index
+ *
+ * @return Old value of the slot
+ *
+ * Complexity: O(1)
+ */
+void *ucs_ptr_array_locked_replace(ucs_ptr_array_locked_t *locked_ptr_array,
+	                           unsigned index, void *new_val);
+
+
+/** Acquire the ptr_array lock
+ *
+ * @param [in] locked_ptr_array  Pointer to a locked ptr array.
+ */
+#define ucs_ptr_array_locked_acquire_lock(_locked_ptr_array) \
+    ucs_recursive_spin_lock(&(_locked_ptr_array)->lock)
+
+
+/** Release the ptr_array lock
+ *
+ * @param [in] locked_ptr_array  Pointer to a locked ptr array.
+ */
+#define ucs_ptr_array_locked_release_lock(_locked_ptr_array) \
+    ucs_recursive_spin_unlock(&(_locked_ptr_array)->lock)
+
+
+/**
+ * Retrieves a value from the locked array.
+ *
+ * The user is required to lock using ucs_ptr_array_locked_acquire_lock()
+ * and ucs_ptr_array_locked_release_lock() functions.
+ *
+ * @param [in]  locked_ptr_array   Pointer to a locked ptr array.
+ * @param [in]  index              Index to retrieve the value from.
+ * @param [out] var                Filled with the value.
+ *
+ * @return Whether the value is present and valid.
+ *
+ * Complexity: O(1)
+ */
+static inline int UCS_F_MAYBE_UNUSED
```

Comment:
why need `UCS_F_MAYBE_UNUSED`?

Reply:
Hi @brminich 

It is needed since it's a static inline function and otherwise it won't build...

Reply:
> It is needed since it's a static inline function and otherwise it won't build...

`static inline` w/o `UCS_F_MAYBE_UNUSED` should be fine
`UCS_F_MAYBE_UNUSED` is needed for `static`-only functions

Reply:
I see, removed it...